### PR TITLE
Add roadmap proposal for incremental sound nnbd

### PR DESCRIPTION
**Pull request by leafpetersen**
@lrhn @eernstg @mit-mit @munificent @vsmenon 

I've been hacking on a writeup of our discussions about an incremental  path to NNBD, and I think it's at a state where some broader feedback would be useful.  I'm putting this up as a PR for now just to make it easy to comment on, but I will probably land it and iterate on it tomorrow.  Please take a look, comments and feedback welcome (on the PR or via email).  This is still very much a high level roadmap, but If we broadly still think this is in line with what we want to propose to do, we should get this out in draft form to leads for discussion next week.

### CODE REVIEW

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 1]
```diff
@@ -0,0 +1,874 @@
...
+Kotlin cannot actually provide full null soundness.
+
+First, Kotlin provides an unsafe cast operator that can be used to escape the type
+system. This primarily arises because Kotlin is implemented over the JVM and
+hence have no way to implement the reification necessary for safe casts at
+composite types.
+```kotlin
+    var nnf : (String) -> Int = { s -> s.length};
+    var nf : (String ?) -> Int = nnf as (String?) -> Int; // Compiler warning.
+    nf(null);  // Will cause a NPE
```
Comment (munificent) [inline working/0110-incremental-sound-nnbd/roadmap.md]: Nit: These code samples don't need to be indented. Here and below.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 2]
```diff
@@ -0,0 +1,874 @@
...
+
+Swift's approach then is to provide a sound type system based around optional
+values, with heavy investment in additional language constructs to make working
+with potentially null values convenient to the programmer, but less investment
+in sophisticated static analysis.  Swift also provides a convenient mechanism
+for falling back to runtime checking with low syntactic overhead where needed.
+
+#### [C#](https://blogs.msdn.microsoft.com/dotnet/2017/11/15/nullable-reference-types-in-csharp/)
+
+C# recently added prototype for non-nullability for default.  Unlike Swift and
```
Comment (munificent) [inline working/0110-incremental-sound-nnbd/roadmap.md]: "prototype" -> "a prototype"

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 3]
```diff
@@ -0,0 +1,874 @@
...
+Kotlin, C# explicitly does not aim for soundess.
```
Comment (munificent) [inline working/0110-incremental-sound-nnbd/roadmap.md]: "soundess" -> "soundness".

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 4]
```diff
@@ -0,0 +1,874 @@
...
+  v : number; // Error, not definitely assigned in constructor
+	constructor(_v: number) { }
+}
+```
+
+As with Kotlin and Swift, Typescript does not seem to have skimped on the static
+analysis work, but instead has invested fairly heavily in providing robust
+checking.  As with C# however, Typescript does seem to have chosen to be unsound
+in a few places out of choice rather than necessity, to relieve the programmer
+of some of the burden of proving safety to the type checker.
```
Comment (munificent) [inline working/0110-incremental-sound-nnbd/roadmap.md]: Probably worth pointing out that even if they were sound, they wouldn't be able to extract any performance benefit from it.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 5]
```diff
@@ -0,0 +1,874 @@
...
+get there.  Specifically, we propose to aim for a system which allows for an
+incremental opt-in migration, and which is fully sound once all code in a
+compilation target has opted-in.
+
+## Motivation for non-nullability in Dart.
+
+This has been fairly well explored previously, so just in brief.  This is one of
+the most requested features in developer surveys, and one of the thing that
+developers most frequently mention missing from other languages.  Kotlin, Swift,
+C#, Typescript, and numerous other languages now have support for
```
Comment (mit-mit) [inline working/0110-incremental-sound-nnbd/roadmap.md]: Isn't C# in the process of adding it for c# 8? Maybe s/now have/now have, or are adding,/ ?

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 6]
```diff
@@ -0,0 +1,874 @@
...
+already opted in (or if it correctly predicts and codes against the
+post-migration API of its dependencies) it only migrates once.
+
+## Proposal roadmap
+
+### Opting in
+
+Non-nullable types will be controlled by a language opt-in as deal with
+elsewhere.  In short, a library may opt-in via syntax in the code, or a package
+may opt-in in entirety.  Opting-in applies only to the library or package so
```
Comment (mit-mit) [inline working/0110-incremental-sound-nnbd/roadmap.md]: I would be great if it was easy to detect when a whole package has opt'ed in. This would make the analysis on Pub much easier.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 7]
```diff
@@ -0,0 +1,874 @@
...
+- The type system should be incrementally sound.  That is, compilation should be
+possible with a mix of opted-in and non-opted-in code, but when all code in a
+compile has opted in, the compiler should be able to trust what the type system
+says about nullability.  When not all code has opted-in, it is possible that a
+non-nullably typed value will receive null.
+- During the migration, packages should be able to opt in without breaking
+  non-opted-in clients.
+- During the migration, packages should be able to opt in before packages that
+  they depend on have opted in.
+- During the migration, if a package opts in after all of its dependencies have
```
Comment (mit-mit) [inline working/0110-incremental-sound-nnbd/roadmap.md]: I found this a bit hard to parse in terms of which direction the dependency is. I think 'all of its dependencies have' means 'all packages it depends on have' ?

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 8]
```diff
@@ -0,0 +1,874 @@
...
+[sound or near-sound checking](https://checkerframework.org/manual/#nullness-checker).
+The annotations are not interpreted with the language at all.  The Java
+documentation
+[suggests](https://blogs.oracle.com/java-platform-group/java-8s-new-type-annotations) not
+relying on nullability annotations, saying "Optional Type Annotations are not a
+substitute for runtime validation".
+
+#### [Typescript](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html)
+
+Typescript is of course unsound by design.  In addition to the core unsoundness
```
Comment (mit-mit) [inline working/0110-incremental-sound-nnbd/roadmap.md]: I'd avoid phrasing this as "of course". I think this is something like "Typescript is unsound by design, as it's layered on top of JavaScripts unsound type system"?

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 9]
```diff
@@ -0,0 +1,874 @@
...
+
+## Summary of proposed goals
+
+We propose to set the following goals for nullability tracking in Dart.
+
+- The type system should be incrementally sound.  That is, compilation should be
+possible with a mix of opted-in and non-opted-in code, but when all code in a
+compile has opted in, the compiler should be able to trust what the type system
+says about nullability.  When not all code has opted-in, it is possible that a
+non-nullably typed value will receive null.
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: Done, thanks.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 10]
```diff
@@ -0,0 +1,874 @@
...
+- During the migration, packages should be able to opt in without breaking
+  non-opted-in clients.
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: Done.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 11]
```diff
@@ -0,0 +1,874 @@
...
+proposals.
+
+### Reification
+
+Nullable (and non-nullable) types are reified, and hence have a runtime effect.
+So for example, casts and instance checks can distinguish between `List<int>`
+and `List<int?>`.  Without runtime reification, nullability is relegated to an
+odd and inconsistent state with respect to the rest of the Dart type system,
+which is uniformly reified.  Null-soundness is also almost certainly impossible
+to achieve in Dart without reification.
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: done.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 12]
```diff
@@ -0,0 +1,874 @@
...
+
+### Soundness
+
+When all libraries in a compile have opted-in to non-nullability, the type
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: done.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 13]
```diff
@@ -0,0 +1,874 @@
...
+system is sound and both compilers and programmers can benefit from a robust
+guarantee that no non-nullably typed value may be observed to be null.  We
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: Added.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 14]
```diff
@@ -0,0 +1,874 @@
...
+dynamic checks so that opted-in code can safely remove null assertions before
+the migration is complete.  Depending on the measured performance impact of
+these checks, we may choose to make these dynamic checks debug mode only.  These
+dynamic checks are orthogonal to any dynamic checking specified as part of the
+core feature, and will not be required when all packages in a compile have opted
+in.  We do not expect that these checks will be sufficient to recover
+null-soundness during the migration period: that is, even with the debug mode
+null guards in place, it will still be possible to see a null pointer error in
+opted-in code because of interactions with non-opted-in code that cannot be
+caught by runtime checks.
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: I don't want to dive too deep into this, but modified the language and added an example to illustrate why we probably don't want to go this route.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 15]
```diff
@@ -0,0 +1,874 @@
...
+
+### Migration path
+
+We propose to support both a waterfall migration (a library may not opt in until
+the transitive closure of its imports have opted-in) and an unconstrained
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: done.

File: working/0110-incremental-sound-nnbd/roadmap.md (hunk 4 above)
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: done.

File: working/0110-incremental-sound-nnbd/roadmap.md (hunk 5 above)
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: done.

File: working/0110-incremental-sound-nnbd/roadmap.md (hunk 6 above)
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: I think should be fairly easy to do.

File: working/0110-incremental-sound-nnbd/roadmap.md (hunk 7 above)
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: done.

File: working/0110-incremental-sound-nnbd/roadmap.md (hunk 8 above)
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: Shortened to "unsound by design".  I don't want go get into why.

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 16]
```diff
@@ -0,0 +1,874 @@
...
+Non-nullable types will be controlled by a language opt-in as deal with
+elsewhere.  In short, a library may opt-in via syntax in the code, or a package
+may opt-in in entirety.  Opting-in applies only to the library or package so
+marked.
+
+Within an opted-in library, unmarked types are interpreted as non-nullable, and
+only types suffixed with `?` are considered nullable.  Additional syntax and
+semantics are enabled for null assertions, definite assignment analysis, type
+promotion.  Additional static errors and warnings are enabled.  If we choose to
+move forward according this roadmap the details of this will be worked out in
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: done

File: working/0110-incremental-sound-nnbd/roadmap.md [hunk 17]
```diff
@@ -0,0 +1,874 @@
...
+other languages.
+
+### Incremental unsoundness and dynamic checking
+
+During the migration period, null-soundness is not guaranteed.  Opted-in code
+that interacts with non-opted-in code may observe null values flowing from a
+non-nullably typed location.  We propose to specify some set of migration period
+dynamic checks so that opted-in code can safely remove null assertions before
+the migration is complete.  Depending on the measured performance impact of
+these checks, we may choose to make these dynamic checks debug mode only.  These
```
[the remark this answers is not in the document]
Comment (leafpetersen) [reply working/0110-incremental-sound-nnbd/roadmap.md]: Either --enable-asserts, or else a specific --enable-null-checking.  Modified the wording to be clearer.

**Comment by leafpetersen** [comment]
Ok, landing this as a draft, and will iterate further in place.
